tests/test_integration.py

```diff
@@ -278,3 +278,44 @@ def test_send_service_called_multiple_times(self):
         response = self.api_msgpack.send_services(services)
         ok_(response)
         eq_(response.status, 200)
+
+@skipUnless(
+    os.environ.get('TEST_DATADOG_INTEGRATION', False),
+    'You should have a running trace agent and set TEST_DATADOG_INTEGRATION=1 env variable'
+)
+class TestAPIDowngrade(TestCase):
+    """
+    Ensures that if the tracing client found an earlier trace agent,
+    it will downgrade the current connection to a stable API version
+    """
+    def test_get_encoder_default(self):
+        # get_encoder should return MsgpackEncoder instance if
+        # msgpack and the CPP implementaiton are available
+        encoder = get_encoder()
+        ok_(isinstance(encoder, MsgpackEncoder))
+
+    @mock.patch('ddtrace.encoding.MSGPACK_ENCODING', False)
+    def test_get_encoder_fallback(self):
+        # get_encoder should return JSONEncoder instance if
+        # msgpack or the CPP implementaiton, are not available
+        encoder = get_encoder()
+        ok_(isinstance(encoder, JSONEncoder))
+
+    def test_downgrade_api(self):
+        # make a call to a not existing endpoint, downgrades
+        # the current API to a stable one
+        tracer = get_dummy_tracer()
+        tracer.trace('client.testing').finish()
+        trace = tracer.writer.pop()
+
+        # the encoder is right but we're targeting an API
+        # endpoint that is not available
+        api = API('localhost', 7777)
+        api._traces = '/v0.0/traces'
+        ok_(isinstance(api._encoder, MsgpackEncoder))
+
+        # after the call, we downgrade to a working endpoint
```

Comment:
after the previous call, we downgrade AND the call should work with a real agent.